x-pack/agent/pkg/agent/application/action_store.go

```diff
@@ -0,0 +1,164 @@
+	reader, err := store.Load()
+	if err != nil {
+		return &actionStore{log: log, store: store}, nil
+	}
+
+	var action actionPolicyChangeSerializer
+
+	dec := yaml.NewDecoder(reader)
+	if err := dec.Decode(&action); err != nil {
+		return nil, err
+	}
+
+	apc := fleetapi.ActionPolicyChange(action)
+
+	return &actionStore{
+		log:    log,
+		store:  store,
+		action: &apc,
+	}, nil
+}
+
+// Add is only taking care of ActionPolicyChange for now and will only keep the last one it receive,
+// any other type of action will be silently ignored.
+func (s *actionStore) Add(a action) {
+	switch v := a.(type) {
+	case *fleetapi.ActionPolicyChange:
+		// Only persist the action if the action is different.
+		if s.action != nil {
+			if s.action.ID() == v.ID() {
+				return
+			}
+		}
+		s.dirty = true
+		s.action = a
+	}
+}
+
+func (s *actionStore) Save() error {
+	defer func() { s.dirty = false }()
+	if !s.dirty {
+		return nil
+	}
+
+	apc, ok := s.action.(*fleetapi.ActionPolicyChange)
+	if !ok {
+		return fmt.Errorf("incompatible type, expected ActionPolicyChange and received %T", s.action)
+	}
+
+	serialize := actionPolicyChangeSerializer(*apc)
+
+	reader, err := yamlToReader(&serialize)
+	if err != nil {
+		return err
+	}
+
+	if err := s.store.Save(reader); err != nil {
+		return err
+	}
+	s.log.Debugf("save on disk action policy change: %+v", s.action)
+	return nil
+}
+
+// Actions returns a slice of action to execute in order, currently only a action policy change is
+// persisted.
+func (s *actionStore) Actions() []action {
+	if s.action == nil {
+		return []action{}
+	}
+
+	return []action{s.action}
+}
+
+// actionPolicyChangeSerializer is a struct that add YAML serialization, I don't think serialization
+// is a concern of the fleetapi package. I went this route so I don't have to do much refactoring.
+//
+// There are four ways to achieve the same results:
+// 1. We create a second struct that map the existing field.
+// 2. We add the serialization in the fleetapi.
+// 3. We move the actual action type outside of the actual fleetapi package.
+// 4. We have two sets of type.
+//
+// This could be done in a refactoring.
+type actionPolicyChangeSerializer struct {
+	ActionID   string                 `yaml:"action_id"`
+	ActionType string                 `yaml:"action_type"`
+	Policy     map[string]interface{} `yaml:"policy"`
+}
+
+// Add a guards between the serializer structs and the original struct.
+var _ actionPolicyChangeSerializer = actionPolicyChangeSerializer(fleetapi.ActionPolicyChange{})
+
+// actionStoreAcker wraps an existing acker and will send any acked event to the action store,
+// its up to the action store to decide if we need to persist the event for future replay or just
+// discard the event.
+type actionStoreAcker struct {
+	acker fleetAcker
+	store *actionStore
+}
+
+func (a *actionStoreAcker) Ack(action fleetapi.Action) error {
+	if err := a.acker.Ack(action); err != nil {
+		return err
+	}
+	a.store.Add(action)
```

Comment:
i'm thinking about scenario when batching is used
policyChange is applied and ACK generated
ACK stored in queue and added into action here
some other action is performed and fails. 
now agent loses both stored state and ACK queue so it is fine i guess.

but if we step aside from batching and each ACK is immediately ACKed as with bare `fleet_acker` then action is ACKed to fleet and stored here into action field.
then agent crashes and action is not stored.

what about (as this does not have a queue of actions but stores only one action) we do store on each Ack
this way we backup latest fully applied configuration regardless of if it was acked or not. 
then on restart it will be acked when applied from backup